lib/vm/src/instance/ref.rs

```diff
@@ -155,3 +155,75 @@ impl InstanceRef {
         (&mut *ptr).as_mut()
     }
 }
+
+/// A weak instance ref. This type does not keep the underlying `Instance` alive
+/// but can be converted into a full `InstanceRef` if the underlying `Instance` hasn't
+/// been deallocated.
+#[derive(Debug, Clone)]
+pub struct WeakInstanceRef(Weak<InstanceInner>);
+
+impl PartialEq for WeakInstanceRef {
+    fn eq(&self, other: &Self) -> bool {
+        self.0.ptr_eq(&other.0)
+    }
+}
+
+impl WeakInstanceRef {
+    /// Try to convert into a strong, `InstanceRef`.
+    pub fn upgrade(&self) -> Option<InstanceRef> {
+        let inner = self.0.upgrade()?;
+        Some(InstanceRef(inner))
+    }
+}
+
+impl MemoryUsage for WeakInstanceRef {
+    fn size_of_val(&self, tracker: &mut dyn MemoryUsageTracker) -> usize {
+        mem::size_of_val(self)
+            + if let Some(ir) = self.upgrade() {
+                ir.size_of_val(tracker)
+            } else {
+                0
+            }
+    }
+}
+
+/// An `InstanceRef` that may or may not be keeping the `Instance` alive.
+///
+/// This type is useful for types that conditionally must keep / not keep the
+/// underlying `Instance` alive. For example, to prevent cycles in `WasmerEnv`s.
+#[derive(Debug, Clone, PartialEq, MemoryUsage)]
+pub enum WeakOrStrongInstanceRef {
+    /// A weak instance ref.
+    Weak(WeakInstanceRef),
+    /// A strong instance ref.
+    Strong(InstanceRef),
+}
+
+impl WeakOrStrongInstanceRef {
+    /// Get a strong `InstanceRef`. A return Value of `None` means that the
+    /// `InstanceRef` has been freed and cannot be accessed.
+    pub fn get_strong(&self) -> Option<InstanceRef> {
+        match self {
+            Self::Weak(weak) => weak.upgrade(),
+            Self::Strong(strong) => Some(strong.clone()),
+        }
+    }
+    /// Get a weak `InstanceRef`.
+    pub fn get_weak(&self) -> WeakInstanceRef {
+        match self {
+            Self::Weak(weak) => weak.clone(),
+            Self::Strong(strong) => WeakInstanceRef(Arc::downgrade(&strong.0)),
+        }
+    }
+
+    /// Convert the existing type directly into a weak reference.
+    pub fn into_weak(&mut self) {
+        let new = self.get_weak();
+        *self = Self::Weak(new);
+    }
+
+    /// Check if the reference contained is strong.
+    pub fn is_strong(&self) -> bool {
+        matches!(self, WeakOrStrongInstanceRef::Strong(_))
+    }
```

Comment:
This is only used in tests. I believe we can move the logic out of this trait (perhaps into the tests?)

Reply:
Yeah, it could be a utility function in tests / in the methods of all the types that use it, it seems like a useful method to have on a type like this though